### DISCUSSION TITLE
nsclient++ on Windows, checking exit_codes

### DISCUSSION THREAD

**Discussion by kuchenmann** [mickem/nscp#816] [General]
I must confess, I do not really understand how this nsclient++ and NRPE are working (Nagios-core).
I have some tasksched_checks, but some of them are show _exit_code'=0;0;0 ' but show CRITICAL-alarms.
Others have _exit_code'=0;0;0 and show "OK".
And I also get exit-codes like _exit_code'=200;0;0 and the show WARNING.
I also get exit code exit_code'=500;0;0 '.

**Comment by mickem** [comment]
Sorry this went unanswered for so long. The confusing part is that the exit code in the performance data and the OK/WARNING/CRITICAL of the check are decided in different places.

**The check status comes from the `warning=` / `critical=` expressions, not from the task's exit code directly.** `check_tasksched` walks every task that matches `filter=`, evaluates the two expressions against each task, and returns the worst result it found. The defaults are:

| Option | Default | Meaning |
|---|---|---|
| `filter` | `enabled = 1` | look at every enabled task |
| `warning` | `exit_code != 0` | any non-zero last result is a WARNING |
| `critical` | `exit_code < 0` | a negative last result is CRITICAL |

That explains your cases:

- `exit_code=200` and `exit_code=500` are non-zero, so they match `exit_code != 0` and give WARNING. Task Scheduler does not interpret these; they are whatever the program the task runs returned.
- A negative code is CRITICAL. `exit_code` is signed 32-bit, so every Windows error HRESULT (`0x8xxxxxxx`, for example `0x80070002` "file not found" or `0x800710E0` "an instance is already running") shows up as a large negative number and trips `exit_code < 0`.
- `'..._exit_code'=0;0;0` inside a CRITICAL result: the check covers *all* enabled tasks unless you add a `filter=`. The perfdata lists one entry per task, so the `=0` entry you looked at is fine, and some *other* task in the same result had a negative code. The message text names the offending task(s); add `show-all` to see every task's line.

**The numbers after the value in the perfdata (`;0;0`) are not thresholds Nagios evaluates.** Nagios never derives a state from performance data; those fields are just the warn/crit values copied out of the two expressions above, for graphing tools. Only the plugin's return code sets the state.

If you want one check per task with your own rules, filter on the task and set the thresholds explicitly. For example, "only a non-zero result is critical, never warn":

```
check_tasksched "filter=title = 'My Backup'" warning=none "critical=exit_code != 0"
```

Or treat your application's own codes the way you mean them:

```
check_tasksched "filter=title = 'My Backup'" "warning=exit_code = 200" "critical=exit_code < 0 OR exit_code >= 500"
```

If a result still does not add up, post the full plugin output (message and perfdata) and the exact command, and I can walk through it.
